{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib as mpl\n",
    "import gffpandas.gffpandas as gffpd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "results_dir = Path(\"/nfs/chisholmlab001/kve/2021_dark_adapted_transcriptome/results/\")\n",
    "\n",
    "experiment_dir = results_dir / \"experiments\"\n",
    "plotting_dir = results_dir / 'gene_expression_plots'\n",
    "\n",
    "log_count_table = experiment_dir / \"experiment_all\" / \"DEseq2out\" / \"experiment_all_NATL2A_rlog.tsv\"\n",
    "rlog_df = pd.read_csv(log_count_table, sep='\\t', index_col=\"long_ID\")\n",
    "\n",
    "rlog_df.columns = pd.MultiIndex.from_product([['Control', 'Pheno'], [0, 4, 8, 13, 16, 20, 24], [1, 2, 3]], names=['treatment', 'time', 'replicate'])\n",
    "\n",
    "rlog_df = rlog_df[~rlog_df.index.duplicated()]\n",
    "\n",
    "rlog_mean_df = rlog_df.groupby(level=['treatment', 'time'], axis='columns').mean()\n",
    "\n",
    "rlog_std_df = rlog_df.groupby(level=['treatment', 'time'], axis='columns').std()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs = []\n",
    "for time in [0, 4, 8, 13, 16, 20, 24]:\n",
    "    result_path = experiment_dir / f\"experiment_{time}\" / \"DGE_tables\" / f\"experiment_{time}_NATL2A_DGE_all.tsv\"\n",
    "    result_df = pd.read_csv(result_path, sep=\"\\t\", index_col=\"long_ID\")\n",
    "\n",
    "    # annotation_data = result_df[[\"product\", \"protein_id\", \"locus_tag\"]]\n",
    "    # annotation_data = annotation_data.drop_duplicates()\n",
    "\n",
    "    significance_data = result_df[[\"log2FoldChange\", \"padj\"]]\n",
    "    significance_data.columns = pd.MultiIndex.from_product([[time], significance_data.columns], names=['time', 'significance'])\n",
    "\n",
    "    significance_data = significance_data.drop_duplicates()\n",
    "\n",
    "    dfs.append(significance_data)\n",
    "\n",
    "result_df = pd.concat(dfs, axis=1)\n",
    "\n",
    "result_df = result_df.swaplevel(\"time\", \"significance\", axis=1)\n",
    "\n",
    "# conversion_df = pd.read_csv(\"gene_label_conversion_table.tsv\", sep='\\t', index_col='gene')\n",
    "\n",
    "# annotation_data = annotation_data.join(conversion_df, how='left')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>Dbxref</th>\n",
       "      <th>...</th>\n",
       "      <th>start_range</th>\n",
       "      <th>strain</th>\n",
       "      <th>transl_table</th>\n",
       "      <th>NCBI ID</th>\n",
       "      <th>NCBI ID_2</th>\n",
       "      <th>NCBI ID_3</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Genbank Annotation</th>\n",
       "      <th>RAST annotation</th>\n",
       "      <th>Response of Synechococcus elongatus PCC7942 homolog to darkness</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cds-WP_011293961.1</th>\n",
       "      <td>NC_007335.2</td>\n",
       "      <td>Protein Homology</td>\n",
       "      <td>CDS</td>\n",
       "      <td>189</td>\n",
       "      <td>1349</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-WP_011293961.1;Parent=gene-PMN2A_RS0001...</td>\n",
       "      <td>Genbank:WP_011293961.1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>11</td>\n",
       "      <td>NATL2_00001</td>\n",
       "      <td>PMN2A_RS00015</td>\n",
       "      <td>PMN2A_1328</td>\n",
       "      <td>dnaN</td>\n",
       "      <td>DNA polymerase III subunit beta</td>\n",
       "      <td>DNA polymerase III beta subunit (EC 2.7.7.7)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-WP_011293962.1</th>\n",
       "      <td>NC_007335.2</td>\n",
       "      <td>Protein Homology</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1352</td>\n",
       "      <td>2122</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-WP_011293962.1;Parent=gene-PMN2A_RS0002...</td>\n",
       "      <td>Genbank:WP_011293962.1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>11</td>\n",
       "      <td>NATL2_00011</td>\n",
       "      <td>PMN2A_RS00020</td>\n",
       "      <td>PMN2A_1329</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>RNA metabolism-related protein</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-WP_011293963.1</th>\n",
       "      <td>NC_007335.2</td>\n",
       "      <td>Protein Homology</td>\n",
       "      <td>CDS</td>\n",
       "      <td>2126</td>\n",
       "      <td>4537</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-WP_011293963.1;Parent=gene-PMN2A_RS0002...</td>\n",
       "      <td>Genbank:WP_011293963.1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>11</td>\n",
       "      <td>NATL2_00021</td>\n",
       "      <td>PMN2A_RS00025</td>\n",
       "      <td>PMN2A_1330</td>\n",
       "      <td>purL</td>\n",
       "      <td>phosphoribosylformylglycinamidine synthase sub...</td>\n",
       "      <td>Phosphoribosylformylglycinamidine synthase, sy...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-WP_011293964.1</th>\n",
       "      <td>NC_007335.2</td>\n",
       "      <td>Protein Homology</td>\n",
       "      <td>CDS</td>\n",
       "      <td>4599</td>\n",
       "      <td>6056</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-WP_011293964.1;Parent=gene-PMN2A_RS0003...</td>\n",
       "      <td>Genbank:WP_011293964.1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>11</td>\n",
       "      <td>NATL2_00031</td>\n",
       "      <td>PMN2A_RS00030</td>\n",
       "      <td>PMN2A_1331</td>\n",
       "      <td>purF</td>\n",
       "      <td>amidophosphoribosyltransferase</td>\n",
       "      <td>Amidophosphoribosyltransferase (EC 2.4.2.14)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds-WP_011293965.1</th>\n",
       "      <td>NC_007335.2</td>\n",
       "      <td>Protein Homology</td>\n",
       "      <td>CDS</td>\n",
       "      <td>6053</td>\n",
       "      <td>8536</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-WP_011293965.1;Parent=gene-PMN2A_RS0003...</td>\n",
       "      <td>Genbank:WP_011293965.1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>11</td>\n",
       "      <td>NATL2_00041</td>\n",
       "      <td>PMN2A_RS00035</td>\n",
       "      <td>PMN2A_1332</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>topoisomerase IV</td>\n",
       "      <td>DNA gyrase subunit A (EC 5.99.1.3)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Yfr22_2</th>\n",
       "      <td>NC_007335.2</td>\n",
       "      <td>source</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>1083668</td>\n",
       "      <td>1083721</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=Yfr22_2;product=Yfr22_2</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Yfr1</th>\n",
       "      <td>NC_007335.2</td>\n",
       "      <td>source</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>1191112</td>\n",
       "      <td>1191170</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=Yfr1;product=Yfr1</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Yfr106</th>\n",
       "      <td>NC_007335.2</td>\n",
       "      <td>source</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>1217566</td>\n",
       "      <td>1217642</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=Yfr106;product=Yfr106</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Yfr23</th>\n",
       "      <td>NC_007335.2</td>\n",
       "      <td>source</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>1084235</td>\n",
       "      <td>1084301</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=Yfr23;product=Yfr23</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AtpT</th>\n",
       "      <td>NC_007335.2</td>\n",
       "      <td>ClaudiaSteglich</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1535062</td>\n",
       "      <td>1535205</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AtpT;Name=Conserved hypothetical protein CD...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1980 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         seq_id            source  type    start      end  \\\n",
       "ID                                                                          \n",
       "cds-WP_011293961.1  NC_007335.2  Protein Homology   CDS      189     1349   \n",
       "cds-WP_011293962.1  NC_007335.2  Protein Homology   CDS     1352     2122   \n",
       "cds-WP_011293963.1  NC_007335.2  Protein Homology   CDS     2126     4537   \n",
       "cds-WP_011293964.1  NC_007335.2  Protein Homology   CDS     4599     6056   \n",
       "cds-WP_011293965.1  NC_007335.2  Protein Homology   CDS     6053     8536   \n",
       "...                         ...               ...   ...      ...      ...   \n",
       "Yfr22_2             NC_007335.2            source  sRNA  1083668  1083721   \n",
       "Yfr1                NC_007335.2            source  sRNA  1191112  1191170   \n",
       "Yfr106              NC_007335.2            source  sRNA  1217566  1217642   \n",
       "Yfr23               NC_007335.2            source  sRNA  1084235  1084301   \n",
       "AtpT                NC_007335.2   ClaudiaSteglich   CDS  1535062  1535205   \n",
       "\n",
       "                   score strand phase  \\\n",
       "ID                                      \n",
       "cds-WP_011293961.1     .      +     0   \n",
       "cds-WP_011293962.1     .      +     0   \n",
       "cds-WP_011293963.1     .      +     0   \n",
       "cds-WP_011293964.1     .      +     0   \n",
       "cds-WP_011293965.1     .      -     0   \n",
       "...                  ...    ...   ...   \n",
       "Yfr22_2                .      -     0   \n",
       "Yfr1                   .      +     0   \n",
       "Yfr106                 .      -     0   \n",
       "Yfr23                  .      +     0   \n",
       "AtpT                   .      +     0   \n",
       "\n",
       "                                                           attributes  \\\n",
       "ID                                                                      \n",
       "cds-WP_011293961.1  ID=cds-WP_011293961.1;Parent=gene-PMN2A_RS0001...   \n",
       "cds-WP_011293962.1  ID=cds-WP_011293962.1;Parent=gene-PMN2A_RS0002...   \n",
       "cds-WP_011293963.1  ID=cds-WP_011293963.1;Parent=gene-PMN2A_RS0002...   \n",
       "cds-WP_011293964.1  ID=cds-WP_011293964.1;Parent=gene-PMN2A_RS0003...   \n",
       "cds-WP_011293965.1  ID=cds-WP_011293965.1;Parent=gene-PMN2A_RS0003...   \n",
       "...                                                               ...   \n",
       "Yfr22_2                                    ID=Yfr22_2;product=Yfr22_2   \n",
       "Yfr1                                             ID=Yfr1;product=Yfr1   \n",
       "Yfr106                                       ID=Yfr106;product=Yfr106   \n",
       "Yfr23                                          ID=Yfr23;product=Yfr23   \n",
       "AtpT                ID=AtpT;Name=Conserved hypothetical protein CD...   \n",
       "\n",
       "                                    Dbxref  ... start_range strain  \\\n",
       "ID                                          ...                      \n",
       "cds-WP_011293961.1  Genbank:WP_011293961.1  ...        None   None   \n",
       "cds-WP_011293962.1  Genbank:WP_011293962.1  ...        None   None   \n",
       "cds-WP_011293963.1  Genbank:WP_011293963.1  ...        None   None   \n",
       "cds-WP_011293964.1  Genbank:WP_011293964.1  ...        None   None   \n",
       "cds-WP_011293965.1  Genbank:WP_011293965.1  ...        None   None   \n",
       "...                                    ...  ...         ...    ...   \n",
       "Yfr22_2                               None  ...        None   None   \n",
       "Yfr1                                  None  ...        None   None   \n",
       "Yfr106                                None  ...        None   None   \n",
       "Yfr23                                 None  ...        None   None   \n",
       "AtpT                                  None  ...        None   None   \n",
       "\n",
       "                   transl_table      NCBI ID      NCBI ID_2   NCBI ID_3  \\\n",
       "ID                                                                        \n",
       "cds-WP_011293961.1           11  NATL2_00001  PMN2A_RS00015  PMN2A_1328   \n",
       "cds-WP_011293962.1           11  NATL2_00011  PMN2A_RS00020  PMN2A_1329   \n",
       "cds-WP_011293963.1           11  NATL2_00021  PMN2A_RS00025  PMN2A_1330   \n",
       "cds-WP_011293964.1           11  NATL2_00031  PMN2A_RS00030  PMN2A_1331   \n",
       "cds-WP_011293965.1           11  NATL2_00041  PMN2A_RS00035  PMN2A_1332   \n",
       "...                         ...          ...            ...         ...   \n",
       "Yfr22_2                    None          NaN            NaN         NaN   \n",
       "Yfr1                       None          NaN            NaN         NaN   \n",
       "Yfr106                     None          NaN            NaN         NaN   \n",
       "Yfr23                      None          NaN            NaN         NaN   \n",
       "AtpT                         11          NaN            NaN         NaN   \n",
       "\n",
       "                   Gene Name  \\\n",
       "ID                             \n",
       "cds-WP_011293961.1      dnaN   \n",
       "cds-WP_011293962.1       NaN   \n",
       "cds-WP_011293963.1      purL   \n",
       "cds-WP_011293964.1      purF   \n",
       "cds-WP_011293965.1      gyrA   \n",
       "...                      ...   \n",
       "Yfr22_2                  NaN   \n",
       "Yfr1                     NaN   \n",
       "Yfr106                   NaN   \n",
       "Yfr23                    NaN   \n",
       "AtpT                     NaN   \n",
       "\n",
       "                                                   Genbank Annotation  \\\n",
       "ID                                                                      \n",
       "cds-WP_011293961.1                    DNA polymerase III subunit beta   \n",
       "cds-WP_011293962.1                               hypothetical protein   \n",
       "cds-WP_011293963.1  phosphoribosylformylglycinamidine synthase sub...   \n",
       "cds-WP_011293964.1                     amidophosphoribosyltransferase   \n",
       "cds-WP_011293965.1                                   topoisomerase IV   \n",
       "...                                                               ...   \n",
       "Yfr22_2                                                           NaN   \n",
       "Yfr1                                                              NaN   \n",
       "Yfr106                                                            NaN   \n",
       "Yfr23                                                             NaN   \n",
       "AtpT                                                              NaN   \n",
       "\n",
       "                                                      RAST annotation  \\\n",
       "ID                                                                      \n",
       "cds-WP_011293961.1       DNA polymerase III beta subunit (EC 2.7.7.7)   \n",
       "cds-WP_011293962.1                     RNA metabolism-related protein   \n",
       "cds-WP_011293963.1  Phosphoribosylformylglycinamidine synthase, sy...   \n",
       "cds-WP_011293964.1       Amidophosphoribosyltransferase (EC 2.4.2.14)   \n",
       "cds-WP_011293965.1                 DNA gyrase subunit A (EC 5.99.1.3)   \n",
       "...                                                               ...   \n",
       "Yfr22_2                                                           NaN   \n",
       "Yfr1                                                              NaN   \n",
       "Yfr106                                                            NaN   \n",
       "Yfr23                                                             NaN   \n",
       "AtpT                                                              NaN   \n",
       "\n",
       "                   Response of Synechococcus elongatus PCC7942 homolog to darkness  \n",
       "ID                                                                                  \n",
       "cds-WP_011293961.1                                                NaN               \n",
       "cds-WP_011293962.1                                                NaN               \n",
       "cds-WP_011293963.1                                                NaN               \n",
       "cds-WP_011293964.1                                                NaN               \n",
       "cds-WP_011293965.1                                                NaN               \n",
       "...                                                               ...               \n",
       "Yfr22_2                                                           NaN               \n",
       "Yfr1                                                              NaN               \n",
       "Yfr106                                                            NaN               \n",
       "Yfr23                                                             NaN               \n",
       "AtpT                                                              NaN               \n",
       "\n",
       "[1980 rows x 45 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_path = Path(\"/nfs/chisholmlab001/kve/2021_dark_adapted_transcriptome/input_data/culture_genome_annotations/NATL2A.gff\")\n",
    "\n",
    "annotation = gffpd.read_gff3(gff_path)\n",
    "annotation_data = annotation.attributes_to_columns()\n",
    "\n",
    "conversion_df = pd.read_csv(\"natl2a_convertion_table2.tsv\", sep='\\t')\n",
    "conversion_df[\"locus_tag\"] = conversion_df['NCBI ID_2']\n",
    "conversion_df = conversion_df[conversion_df[\"locus_tag\"].notna()]\n",
    "\n",
    "annotation_data = annotation_data.merge(conversion_df, how='left', on=\"locus_tag\")\n",
    "annotation_data = annotation_data.set_index(\"ID\")\n",
    "\n",
    "annotation_data = annotation_data[annotation_data['type'].isin(['sRNA', 'CDS'])]\n",
    "annotation_data = annotation_data.drop_duplicates()\n",
    "\n",
    "annotation_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "@mpl.rc_context({\n",
    "    'lines.linewidth': 6, \n",
    "    'lines.marker':'o', \n",
    "    'lines.markersize':18, \n",
    "    'legend.fontsize': 'x-large',\n",
    "    'axes.labelsize': 'x-large',\n",
    "    'axes.titlesize':'x-large',\n",
    "    'xtick.labelsize':'x-large',\n",
    "    'ytick.labelsize':'x-large'})\n",
    "def plot_gene(ax, gene_ID, rlog_mean_df, rlog_std_df, results_df, annotation_data, night_periods, night_color, attr_dict):\n",
    "\n",
    "    ax.yaxis.set_major_locator(mpl.ticker.MaxNLocator(integer=True))\n",
    "    mean_series = rlog_mean_df.loc[gene_ID]\n",
    "    std_series = rlog_std_df.loc[gene_ID]\n",
    "\n",
    "    for treatment in ['Control', 'Pheno']:\n",
    "        for (s, e) in night_periods:\n",
    "            ax.axvspan(s, e, color=night_color)\n",
    "        ax.errorbar(x=mean_series[treatment].index, y=mean_series[treatment], yerr=std_series[treatment], capsize=6, capthick=3, label=treatment, color=attr_dict[treatment]['color'])\n",
    "    \n",
    "    # axes limits\n",
    "    bottom, top = ax.get_ylim()\n",
    "    y_range = top - bottom\n",
    "    ax.set_ylim(top-(y_range*1.1), top)\n",
    "    ax.set_xlim(-1, 25)\n",
    "    \n",
    "    #plotting significance\n",
    "    significance_df = result_df.loc[gene_ID]\n",
    "    significance_df[\"padj\"]\n",
    "\n",
    "    for x, padj in zip(significance_df[\"padj\"].index, significance_df[\"padj\"]):\n",
    "        if padj < 0.05:\n",
    "            ax.plot(x, bottom - y_range*0.05, color='k', marker=(8, 2, 0), markersize=15, label=\"Differentially expressed at 5% FDR\")\n",
    "\n",
    "    gene_annotation = annotation_data.loc[gene_ID]\n",
    "\n",
    "    # title stuff\n",
    "    ax.set_title(f\"{gene_annotation.name}/{gene_annotation['NCBI ID_3']}\\n{gene_annotation['Genbank Annotation']}\")\n",
    "\n",
    "@mpl.rc_context({\n",
    "    'lines.linewidth': 6, \n",
    "    'lines.marker':'o', \n",
    "    'lines.markersize':18, \n",
    "    'legend.fontsize': 'x-large',\n",
    "    'axes.labelsize': 'xx-large',\n",
    "    'axes.titlesize':'xx-large',\n",
    "    'xtick.labelsize':'xx-large',\n",
    "    'figure.titlesize': 'xx-large',\n",
    "    'ytick.labelsize':'xx-large'})\n",
    "def plot_gene_table(gene_df_subset, out_path, rlog_mean_df, rlog_std_df, results_df, annotation_data,\n",
    "                num_cols = 3,\n",
    "                night_periods = [(-11, 0), (13, 24)], \n",
    "                night_color=\"#dfdfdf\",\n",
    "                attr_dict={'Control':{'color':'salmon', 'label':'Parental $\\it{Prochlorococcus}$'}, 'Pheno':{'color':'lightseagreen', 'label':'Dark-tolerant $\\it{Prochlorococcus}$'}}):\n",
    "\n",
    "    # original from elaina:\n",
    "    # color_dict={'Control':'#e97e72', 'Pheno':'#52bcc2'}\n",
    "\n",
    "    if len(gene_df_subset) > num_cols:\n",
    "        gene_arr = list(gene_df_subset.index.values)\n",
    "        gene_arr += [None]*(num_cols - (len(gene_arr) % num_cols))\n",
    "        gene_arr = np.array(gene_arr).reshape(-1, num_cols)\n",
    "    else:\n",
    "        gene_arr = np.array(gene_df_subset.index.values).reshape(1, len(gene_df_subset))\n",
    "\n",
    "    y_height = 5\n",
    "    x_width = 5\n",
    "\n",
    "    heights = [y_height]*gene_arr.shape[0]\n",
    "    widths = [x_width]*gene_arr.shape[1]\n",
    "\n",
    "    fig = plt.figure(figsize=(sum(widths), sum(heights)), constrained_layout=True)\n",
    "    gs = fig.add_gridspec(ncols=len(widths), nrows=len(heights), height_ratios=heights, width_ratios=widths)\n",
    "\n",
    "    for i, row in enumerate(gene_arr):\n",
    "        for j, element in enumerate(row):\n",
    "            if element != None:\n",
    "                ax = fig.add_subplot(gs[i,j])\n",
    "                plot_gene(ax, element, rlog_mean_df, rlog_std_df, results_df, annotation_data, night_periods, night_color, attr_dict)\n",
    "\n",
    "    # handles, labels = ax.get_legend_handles_labels()\n",
    "    legend_elements = [mpl.lines.Line2D([0], [0], color=d['color'], label=d['label']) for t, d in attr_dict.items()]\n",
    "    legend = fig.legend(handles=legend_elements, loc='center left', bbox_to_anchor= (1.01, 0.5))\n",
    "\n",
    "    xlab = fig.supxlabel(\"Time (hours)\")\n",
    "    ylab = fig.supylabel(\"Relative transcript abundance\")\n",
    "    plt.savefig(out_path, bbox_extra_artists=[legend, xlab, ylab], bbox_inches='tight')\n",
    "    plt.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clock proteins\n",
    "clock_proteins = [\"KaiB\",\n",
    "                  \"KaiC\",\n",
    "                  \"SasA\",\n",
    "                  \"RpaA\",\n",
    "                  \"LdpA\"]\n",
    "\n",
    "clock_proteins = [\"PMN2A_0914\", \"PMN2A_0913\", \"PMN2A_0674\", \"PMN2A_1494\", \"PMN2A_1131\"]\n",
    "\n",
    "genes = annotation_data[annotation_data[\"NCBI ID_3\"].isin(clock_proteins)]\n",
    "\n",
    "out_path = plotting_dir / 'clock_proteins.jpg'\n",
    "plot_gene_table(genes, out_path, rlog_mean_df, rlog_std_df, result_df, annotation_data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sigma Factors\n",
    "sigma_factors = [\"PMN2A_RS02610\",\n",
    "                 \"PMN2A_RS03005\",\n",
    "                 \"PMN2A_RS07470\",\n",
    "                 \"PMN2A_RS09305\",\n",
    "                 \"PMN2A_RS09820\"]\n",
    "\n",
    "genes = annotation_data[annotation_data[\"NCBI ID_3\"].isin(clock_proteins)]\n",
    "\n",
    "out_path = plotting_dir / 'clock_proteins.jpg'\n",
    "plot_gene_table(genes, out_path, rlog_mean_df, rlog_std_df, result_df, annotation_data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stringent Reponse Genes\n",
    "stringent_response = [\"PMN2A_1459\", \"PMN2A_1467\", \"PMN2A_1474\", \"PMN2A_1490\", \"PMN2A_1506\", \"PMN2A_1515\", \"PMN2A_1541\", \"PMN2A_1580\", \"PMN2A_1599\", \"PMN2A_1607\", \"PMN2A_1646\", \"PMN2A_1661\", \"PMN2A_1721\", \"PMN2A_1735\", \"PMN2A_1738\", \"PMN2A_1776\", \"PMN2A_1777\", \"PMN2A_1778\", \"PMN2A_1779\", \"PMN2A_1879\", \"PMN2A_1880\", \"PMN2A_0016\", \"PMN2A_0066\", \"PMN2A_0183\", \"PMN2A_0437\", \"PMN2A_0621\", \"PMN2A_0735\", \"PMN2A_0758\", \"PMN2A_0830\", \"PMN2A_0847\", \"PMN2A_0975\", \"PMN2A_0995\", \"PMN2A_1005\", \"PMN2A_1075\", \"PMN2A_1130\", \"PMN2A_1144\", \"PMN2A_1150\", \"PMN2A_1190\", \"PMN2A_1311\", \"PMN2A_1326\", \"PMN2A_0077\", \"PMN2A_1217\"]\n",
    "\n",
    "genes = annotation_data[annotation_data[\"NCBI ID_3\"].isin(stringent_response)]\n",
    "out_path = plotting_dir / 'stringent_response.jpg'\n",
    "plot_gene_table(genes, out_path, rlog_mean_df, rlog_std_df, result_df, annotation_data, num_cols=7)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# histidine kinases and response regulator genes\n",
    "histidine_kinase_w_response_regulators = [\"PMN2A_0437\",\n",
    "                                          \"PMN2A_0674\",\n",
    "                                          \"PMN2A_0912\",\n",
    "                                          \"PMN2A_1151\",\n",
    "                                          \"PMN2A_1635\",\n",
    "                                          \"PMN2A_0436\",\n",
    "                                          \"PMN2A_0637\",\n",
    "                                          \"PMN2A_1196\",\n",
    "                                          \"PMN2A_1494\",\n",
    "                                          \"PMN2A_1500\",\n",
    "                                          \"PMN2A_1537\",]\n",
    "\n",
    "genes = annotation_data[annotation_data[\"NCBI ID_3\"].isin(histidine_kinase_w_response_regulators)]\n",
    "out_path = plotting_dir / 'histidine_kinase_w_response_regulators.jpg'\n",
    "plot_gene_table(genes, out_path, rlog_mean_df, rlog_std_df, result_df, annotation_data, num_cols=4)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# atpT (PMN2A_2168) - cyanobacterial ATP synthase inhibitory factor\n",
    "\n",
    "genes = annotation_data[annotation_data[\"locus_tag\"].isin([\"PMN2A_2168\"])]\n",
    "out_path = plotting_dir / 'atpT.jpg'\n",
    "plot_gene_table(genes, out_path, rlog_mean_df, rlog_std_df, result_df, annotation_data, num_cols=4)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ATP Synthase Gene Cluster -- Product contains \"ATP synthase\" -- should be 9 genes\n",
    "\n",
    "genes = annotation_data[annotation_data[\"product\"].str.contains(\"ATP synthase\")]\n",
    "out_path = plotting_dir / 'ATPsynthaseSubunits.jpg'\n",
    "plot_gene_table(genes, out_path, rlog_mean_df, rlog_std_df, result_df, annotation_data, num_cols=4)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Transcriptional repressor LexA (PMN2A_0828)\n",
    "genes = annotation_data[annotation_data[\"NCBI ID_3\"].isin([\"PMN2A_0828\"])]\n",
    "out_path = plotting_dir / 'LexA.jpg'\n",
    "plot_gene_table(genes, out_path, rlog_mean_df, rlog_std_df, result_df, annotation_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pcb gene family cluster\n",
    "# get from Allison's most recent email\n",
    "# It should have PMN2A_:\n",
    "# PMN2A_0066 pcbA\n",
    "# PMN2A_0719 pcbB1\n",
    "# PMN2A_0723 pcbB2 (which is what we are missing)\n",
    "# PMN2A_0215 pcbC\n",
    "# PMN2A_0718 pcbD\n",
    "# PMN2A_0722 pcbE\n",
    "# PMN2A_0717 pcbH\n",
    "\n",
    "pcb = [\"PMN2A_0215\", \"PMN2A_0066\", \"PMN2A_0717\", \"PMN2A_0718\", \"PMN2A_0719\", \"PMN2A_0723\", \"PMN2A_0722\"]\n",
    "genes = annotation_data[annotation_data[\"NCBI ID_3\"].isin(pcb)]\n",
    "out_path = plotting_dir / 'pcb.jpg'\n",
    "plot_gene_table(genes, out_path, rlog_mean_df, rlog_std_df, result_df, annotation_data, num_cols=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Transcriptional regulator AbrB (PMN2A_1774)\n",
    "genes = annotation_data[annotation_data[\"NCBI ID_3\"].isin([\"PMN2A_1774\"])]\n",
    "out_path = plotting_dir / 'AbrB.jpg'\n",
    "plot_gene_table(genes, out_path, rlog_mean_df, rlog_std_df, result_df, annotation_data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# two component transcriptional regulator (PMN2A_0637/PMN2A_RS06350)\n",
    "genes = annotation_data[annotation_data[\"NCBI ID_3\"].isin([\"PMN2A_0637\"])]\n",
    "out_path = plotting_dir / 'response_regulator_transcription_factor.jpg'\n",
    "plot_gene_table(genes, out_path, rlog_mean_df, rlog_std_df, result_df, annotation_data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# transcriptional regulator, GntR family (PMN2A_1216)\n",
    "genes = annotation_data[annotation_data[\"NCBI ID_3\"].isin([\"PMN2A_1216\"])]\n",
    "out_path = plotting_dir / 'GntR_family_transcriptional_regulator.jpg'\n",
    "plot_gene_table(genes, out_path, rlog_mean_df, rlog_std_df, result_df, annotation_data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# YFRs\n",
    "# genes = annotation_data[annotation_data[\"NCBI ID_3\"].isin([\"PMN2A_1216\"])]\n",
    "\n",
    "# YFR103 specifically\n",
    "\n",
    "index_series = annotation_data.index.to_series()\n",
    "yfr_list = index_series[index_series.str.contains(\"Yfr\")==True].to_list()\n",
    "\n",
    "genes = annotation_data.loc[yfr_list]\n",
    "out_path = plotting_dir / 'yfrs.jpg'\n",
    "plot_gene_table(genes, out_path, rlog_mean_df, rlog_std_df, result_df, annotation_data, num_cols=5)\n",
    "\n",
    "genes = annotation_data.loc[['Yfr103_1', 'Yfr103_2']]\n",
    "out_path = plotting_dir / 'yfr103.jpg'\n",
    "plot_gene_table(genes, out_path, rlog_mean_df, rlog_std_df, result_df, annotation_data)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sigma 70 – rpoD (PMN2A_1829)\n",
    "genes = annotation_data[annotation_data[\"NCBI ID_3\"].isin([\"PMN2A_1829\"])]\n",
    "out_path = plotting_dir / 'sigma70.jpg'\n",
    "plot_gene_table(genes, out_path, rlog_mean_df, rlog_std_df, result_df, annotation_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "48dc67bfef3c61a86793ed8beb21736f35fa3d23295e6a77f8d0bde0ce0b5d2f"
  },
  "kernelspec": {
   "display_name": "Python 3.9.7 64-bit ('base': conda)",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
